sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/StartingPointShardsFinder.java

```diff
@@ -0,0 +1,172 @@
+   *            /
+   * 0001------+
+   *            \
+   *             0004-----------+             0008------------------
+   *                             \           /
+   *                              0007------+
+   *                             /           \
+   *                  0005------+             0009------+
+   *                 /                                   \
+   * 0002-----------+                                     0010------
+   *                 \                                   /
+   *                  0006------------------------------+
+   * </pre>
+   */
+  Iterable<Shard> findShardsAtStartingPoint(SimplifiedKinesisClient kinesis, String streamName,
+      StartingPoint startingPoint) throws TransientKinesisException {
+    List<Shard> allShards = kinesis.listShards(streamName);
+    Set<Shard> initialShards = findInitialShardsWithoutParents(streamName, allShards);
+
+    Set<Shard> startingPointShards = new HashSet<>();
+    Sets.SetView<Shard> expiredShards;
+    do {
+      Set<Shard> validShards = validateShards(kinesis, initialShards, streamName, startingPoint);
+      startingPointShards.addAll(validShards);
+      expiredShards = Sets.difference(initialShards, validShards);
+      Set<String> expiredShardIds = new HashSet<>();
+      for (Shard expiredShard : expiredShards) {
+        expiredShardIds.add(expiredShard.getShardId());
+      }
+      if (!expiredShardIds.isEmpty()) {
+        LOGGER.info("Following shards expired for {} stream at '{}' starting point: {}", streamName,
+            startingPoint, expiredShardIds);
+      }
+      initialShards = new HashSet<>();
+      for (Shard shard : allShards) {
+        if (expiredShardIds.contains(shard.getParentShardId())) {
+          initialShards.add(shard);
+        }
+      }
+    } while (!expiredShards.isEmpty());
+    return startingPointShards;
+  }
+
+  /**
+   * Finds the initial set of shards (the oldest ones). These shards do not have their parents in
+   * the shard list.
+   */
+  private Set<Shard> findInitialShardsWithoutParents(String streamName, List<Shard> allShards) {
+    Set<String> shardIds = new HashSet<>();
+    for (Shard shard : allShards) {
+      shardIds.add(shard.getShardId());
+    }
+    LOGGER.info("Stream {} has following shards: {}", streamName, shardIds);
+    Set<Shard> shardsWithoutParents = new HashSet<>();
+    for (Shard shard : allShards) {
+      if (!shardIds.contains(shard.getParentShardId())) {
+        shardsWithoutParents.add(shard);
+      }
+    }
+    return shardsWithoutParents;
+  }
+
+  private Set<Shard> validateShards(SimplifiedKinesisClient kinesis, Iterable<Shard> rootShards,
+      String streamName, StartingPoint startingPoint)
+      throws TransientKinesisException {
+    Set<Shard> validShards = new HashSet<>();
+    ShardIteratorType shardIteratorType = ShardIteratorType
+        .fromValue(startingPoint.getPositionName());
+    for (Shard shard : rootShards) {
```

Comment:
How many can there be? Is it a concern that we'll be sequentially making this many getRecords calls?

Reply:
There's a limit of 500 shards per AWS region (or 200 shards in some regions). More on limits here: https://docs.aws.amazon.com/streams/latest/dev/service-sizes-and-limits.html. So it might be a problem in extreme cases only.